admittance_controller/include/admittance_controller/admittance_rule.hpp

```diff
@@ -0,0 +1,191 @@
+
+  class AdmittanceRule {
+  public:
+    AdmittanceRule() {
+      parameters_ = &parameters_copy;
+    }
+
+    controller_interface::return_type
+    configure(const std::shared_ptr<rclcpp_lifecycle::LifecycleNode> &node, int num_joint);
+
+    controller_interface::return_type reset();
+
+    /**
+     * Calculate 'desired joint states' based on the 'measured force' and 'reference joint state'.
+     *
+     * \param[in] current_joint_state
+     * \param[in] measured_wrench
+     * \param[in] reference_joint_state
+     * \param[in] period
+     * \param[out] desired_joint_state
+     */
+    controller_interface::return_type update(
+        const trajectory_msgs::msg::JointTrajectoryPoint &current_joint_state,
+        const geometry_msgs::msg::Wrench &measured_wrench,
+        const trajectory_msgs::msg::JointTrajectoryPoint &reference_joint_state,
+        const rclcpp::Duration &period,
+        trajectory_msgs::msg::JointTrajectoryPoint &desired_joint_states);
+
+    void get_controller_state(control_msgs::msg::AdmittanceControllerState &state_message);
+
+
+  public:
+    // admittance config parameters
+    std::shared_ptr<admittance_struct_parameters::admittance_struct> parameter_handler_;
+    admittance_struct_parameters::admittance_struct::params *parameters_;
+    admittance_struct_parameters::admittance_struct::params parameters_copy;
+
+  protected:
+    /**
+     * All values are in the controller frame
+     */
+    void calculate_admittance_rule(
+        const Eigen::Matrix<double, 3, 2> &wrench,
+        const Eigen::Matrix<double, 3, 2> &desired_vel,
+        const double dt
+    );
+
+    void process_wrench_measurements(
+        const geometry_msgs::msg::Wrench &measured_wrench, const Eigen::Matrix<double, 3, 3> &sensor_rot,
+        const Eigen::Matrix<double, 3, 3> &cog_rot
+    );
+
+    Eigen::Vector3d get_rotation_axis(const Eigen::Matrix3d &R) const;
+
+    void convert_cartesian_deltas_to_joint_deltas(const std::vector<double> &positions,
+                                                  const Eigen::Matrix<double, 3, 2> &cartesian_delta,
+                                                  const std::string &link_name,
+                                                  Eigen::Matrix<double, Eigen::Dynamic, 1> &joint_delta, bool &success);
+
+    Eigen::Matrix<double, 3, 2> convert_joint_deltas_to_cartesian_deltas(const std::vector<double> &positions,
+                                                                         const std::vector<double> &joint_delta,
+                                                                         const std::string &link_name,
+                                                                         bool &success);
+
+    void normalize_rotation(Eigen::Matrix<double, 3, 3, Eigen::ColMajor> &R);
+
+    Eigen::Matrix<double, 4, 4, Eigen::ColMajor>
+    get_transform(const std::vector<double> &positions, const std::string &link_name, bool external, bool &success);
```

Comment:
This is a wrapper method around the kinematics interface. Either the  API for kinematics interface can change or we have to use a wrapper. Although, it might not be a bad idea to move the TF logic into the kinematics interface.